language/stdlib/modules/Diem.move

```diff
@@ -556,36 +619,66 @@ module Diem {
         include BurnWithResourceCapEnsures<CoinType>{preburn: global<Preburn<CoinType>>(preburn_address)};
     }
 
+
+    // resource struct ConcurrentPreburns<CoinType> {
+    //     inflight_burns: vector<Preburn<CoinType>>,
+    // }
+    // resource struct Preburn<CoinType> {
+    //     /// A single pending burn amount.
+    //     /// There is no pending burn request if the value in `to_burn` is 0
+    //     to_burn: Diem<CoinType>,
+    // }
+
     /// Permanently removes the coins held in the `Preburn` resource (in `to_burn` field)
     /// stored at `preburn_address` and updates the market cap accordingly.
     /// This function can only be called by the holder of a `BurnCapability<CoinType>`.
     /// Calls to this function will fail if the there is no `Preburn<CoinType>`
     /// resource under `preburn_address`, or, if the preburn `to_burn` area for
     /// `CoinType` is empty (error code 7).
     fun burn_with_resource_cap<CoinType>(
-        preburn: &mut Preburn<CoinType>,
+        concurrent_preburns: &mut ConcurrentPreburns<CoinType>,
         preburn_address: address,
+        amount_to_burn: u64,
         _capability: &BurnCapability<CoinType>
     ) acquires CurrencyInfo {
         let currency_code = currency_code<CoinType>();
-        // Abort if no coin present in preburn area
-        assert(preburn.to_burn.value > 0, Errors::invalid_state(EPREBURN_EMPTY));
-        // destroy the coin in Preburn area
-        let Diem { value } = withdraw_all<CoinType>(&mut preburn.to_burn);
-        // update the market cap
         assert_is_currency<CoinType>();
+
+        // Abort if no coin present in preburn queue
+
+        assert(len(concurrent_preburns.inflight_burns.value) == 0, Errors::invalid_state(EPREBURN_EMPTY));
+
+        let number_preburns_inflight = Vector::length(concurrent_preburns.inflight_burns);
+        let preburn_index = 0;
+
+        while (preburn_index < number_preburns_inflight) {
+            let burn = *Vector::borrow(concurrent_preburns.inflight_burns, preburn_index);
+            let burn_amount = burn.to_burn.amount;
+            if burn_amount == amount {
+                let to_burn = Vector::remove<Preburn<CoinType>>(concurrent_preburns.inflight_burns, index);
+                // destroy the coin
+                let Diem { _value } = withdraw_all<CoinType>(&mut to_burn.to_burn);
```

Comment:
actually return to_burn instead 'return to_burn' (no break).

then abort error instead of return